maia/src/protocol.rs

```diff
@@ -437,48 +334,51 @@ impl Payout {
 
         txouts
     }
+}
 
-    /// Subtracts fee fairly from both outputs
-    ///
-    /// We need to consider a few cases:
-    /// - If both amounts are >= DUST, they share the fee equally
-    /// - If one amount is < DUST, it set to 0 and the other output needs to cover for the fee.
-    fn with_updated_fee(
-        self,
-        fee: Amount,
-        dust_limit_maker: Amount,
-        dust_limit_taker: Amount,
-    ) -> Result<Self> {
-        let maker_amount = self.maker_amount;
-        let taker_amount = self.taker_amount;
-
-        let mut updated = self;
-        match (
-            maker_amount
-                .checked_sub(fee / 2)
-                .map(|a| a > dust_limit_maker)
-                .unwrap_or(false),
-            taker_amount
-                .checked_sub(fee / 2)
-                .map(|a| a > dust_limit_taker)
-                .unwrap_or(false),
-        ) {
-            (true, true) => {
-                updated.maker_amount -= fee / 2;
-                updated.taker_amount -= fee / 2;
-            }
-            (false, true) => {
-                updated.maker_amount = Amount::ZERO;
-                updated.taker_amount = taker_amount - (fee + maker_amount);
-            }
-            (true, false) => {
-                updated.maker_amount = maker_amount - (fee + taker_amount);
-                updated.taker_amount = Amount::ZERO;
-            }
-            (false, false) => bail!("Amounts are too small, could not subtract fee."),
+/// Subtracts fee fairly from both outputs
+///
+/// We need to consider a few cases:
+/// - If both amounts are >= DUST, they share the fee equally
+/// - If one amount is < DUST, it set to 0 and the other output needs to cover for the fee.
+fn update_payout_fee(
```

Comment:
❓ With the library split into two crates, how do we decide what goes where?

Obviously this change is motivated by a bug which we need to deal with somehow, but I don't see why this code you left here belongs in `maia` over `maia-core`, other than because we need to make changes to it and depend on two versions of it in `itchysats`.

The separation seems arbitrary to me in this case. I wonder if we should be doing something else that doesn't require introducing these changes.

Reply:
My primary motivation was to move to `maia-core` the following things:
- types, without the behaviour/policy (only moved `generate_payouts()` factory function there as well, as it's essentially like a constructor, and needs access to private fields
- basic utilities (like `Digits` etc)

Please note that `update_payout_fee` is specifying a policy (ie. how we split the fees), and I feel that is something that should not go to `maia-core`. I think it is consistent with the recent discussion with @thomaseizinger about separating policy code.

Reply:
Okay, I like the motivation of separating policy from mechanism. It does allow us to change the policy without necessarily introducing breaking changes in `itchysats`.

I can also see how `update_payout_fee` is specifying a policy.

I do think that it is not so straightforward to decide where different things belong, so we will need to be careful going forward.

Reply:
`maia` is in many ways a policy-free crate already. That was one of the design decisions of just exporting free functions (and not full state machines like we did in the past). I am not sure what the bug at hand is but perhaps exposing more configuration knobs / parameters can also solve the issue?

The data structures in maia purposely try to not be "clever" and don't employ any information-hiding because we only export free functions and thus all the structs are really just "bags of data" that we pass in and out.

Reply:
> `maia` is in many ways a policy-free crate already. That was one of the design decisions of just exporting free functions (and not full state machines like we did in the past). I am not sure what the bug at hand is but perhaps exposing more configuration knobs / parameters can also solve the issue?
> 
> The data structures in maia purposely try to not be "clever" and don't employ any information-hiding because we only export free functions and thus all the structs are really just "bags of data" that we pass in and out.

#34 is sorting out a bug - there are some parts of policy unfortunately, especially the `update_payout_fee` as this one specifies *how* to split the fees. 

I must say that the only reason why we could split the crate so easily in the first place was because it was designed well!

This is an interesting point raised by you - we could have added an enum about splitting the prices in two different ways. This would be less ugly solution then duplicating the API (but would allow people to choose the "unfair" fair split in this case). I will update the other comment as it's worth mentioning.

Reply:
An enum allows you to choose between policies. That doesn't make the code policy-free.

A closure (or in design-pattern language: strategy pattern), would allow the caller to pass in a fee-splitting policy as a value without maia ever having to know about the specifics.